runtime/compiler/optimizer/MethodHandleTransformer.cpp

```diff
@@ -0,0 +1,562 @@
+      LocalObjectInfo firstObj = (*first)[i];
+      LocalObjectInfo secondObj = (*second)[i];
+
+      if (firstObj == UNUSED)
+         {
+         (*first)[i] = secondObj;
+         }
+      else if (secondObj != UNUSED && firstObj != secondObj)
+         {
+         (*first)[i] = UNKNOWN;
+         }
+
+      if (firstObj != (*first)[i])
+         changed = true;
+      }
+
+   if (trace())
+      {
+      if (changed)
+         {
+         traceMsg(comp(), "Object info after merging:\n");
+         printObjectInfo(first, comp());
+         }
+      else
+         traceMsg(comp(), "Object info is not changed after merging\n");
+      }
+   }
+
+// Given a address type node, try to retrieve or compute its value
+//
+TR_MethodHandleTransformer::LocalObjectInfo TR_MethodHandleTransformer::getObjectInfoOfNode(TR::Node* node)
+   {
+   TR_ASSERT(node->getType() == TR::Address, "Can't have object info on non-address type node n%dn %p", node->getGlobalIndex(), node);
+
+   if (trace())
+      {
+      traceMsg(comp(), "Looking for object info of n%dn\n", node->getGlobalIndex());
+      if (_currentObjectInfo)
+         {
+         traceMsg(comp(), "Current object info is:\n");
+         printObjectInfo(_currentObjectInfo, comp());
+         }
+      }
+
+   if (!node->getOpCode().hasSymbolReference())
+      return UNKNOWN;
+
+   auto symRef = node->getSymbolReference();
+   auto symbol = symRef->getSymbol();
+
+   if (symRef->isUnresolved())
+      return UNKNOWN;
+
+   if (symRef->hasKnownObjectIndex())
+      return symRef->getKnownObjectIndex();
+
+   auto knot = comp()->getKnownObjectTable();
+
+   if (node->getOpCode().isLoadDirect() &&
+       symbol->isAutoOrParm())
+      {
+      traceMsg(comp(), "getObjectInfoOfNode n%dn is load from auto or parm, local #%d\n", node->getGlobalIndex(), symbol->getLocalIndex());
+      return _currentObjectInfo ? (*_currentObjectInfo)[symbol->getLocalIndex()] : UNKNOWN;
+      }
+
+   if (knot &&
+       node->getOpCode().isCall() &&
+       !symbol->castToMethodSymbol()->isHelper())
+      {
+      auto rm = symbol->castToMethodSymbol()->getMandatoryRecognizedMethod();
+      switch (rm)
+        {
+        case TR::java_lang_invoke_DirectMethodHandle_internalMemberName:
+        case TR::java_lang_invoke_DirectMethodHandle_internalMemberNameEnsureInit:
+           {
+           auto mhIndex = getObjectInfoOfNode(node->getFirstArgument());
+           if (isKnownObject(mhIndex) && !knot->isNull(mhIndex))
+              {
+              TR::VMAccessCriticalSection dereferenceKnownObjectField(comp()->fej9());
+              uintptr_t mhObject = knot->getPointer(mhIndex);
+              uintptr_t mnObject = comp()->fej9()->getReferenceField(mhObject, "member", "Ljava/lang/invoke/MemberName;");
+              auto mnIndex = knot->getOrCreateIndex(mnObject);
+              if (trace())
+                 traceMsg(comp(), "Get DirectMethodHandle.member known object %d\n", mnIndex);
+              return mnIndex;
```

Comment:
@dmitry-ten This sequence and the one below will need a message to the JITServer client

Reply:
Yeah, I will change this to support JITServer in a separate PR. 
As I understand, OpenJDK method handles are not yet enabled by default, so we can leave it like this for now.